compiler/catala_utils/messages.ml

```diff
@@ -2,188 +2,232 @@
+  Ocolor_format.prettify_formatter ppf;
+  ppf
+
+(* Sets handling of tags in the formatter to ignore them (don't print any color
+   codes) *)
+let unstyle_formatter ppf =
+  Format.pp_set_mark_tags ppf false;
+  ppf
+(* Format.pp_set_formatter_stag_functions ppf {
+ *   Format.mark_open_stag = (fun _ -> "");
+ *   mark_close_stag = (fun _ -> "");
+ *   print_open_stag = ignore;
+ *   print_close_stag = ignore;
+ * };
+ * ppf *)
+
+(* SIDE EFFECT AT MODULE LOAD: this turns on handling of tags in
+   [Format.sprintf] etc. functions (ignoring them) *)
+let () = ignore (unstyle_formatter Format.str_formatter)
+
+(* Note: we could do the same for std_formatter, err_formatter... but we'd
+   rather promote the use of the formatting functions of this module and the
+   below std_ppf / err_ppf *)
+
+let has_color oc =
+  match !Cli.style_flag with
+  | Cli.Never -> false
+  | Always -> true
+  | Auto -> Unix.(isatty (descr_of_out_channel oc))
+
+(* Here we create new formatters to stderr/stdout that remain separate from the
+   ones used by [Format.printf] / [Format.eprintf] (which remain unchanged) *)
+
+let formatter_of_out_channel oc =
+  let ppf = Format.formatter_of_out_channel oc in
+  if has_color oc then color_formatter ppf else unstyle_formatter ppf
+
+let std_ppf = lazy (formatter_of_out_channel stdout)
+let err_ppf = lazy (formatter_of_out_channel stderr)
+let ignore_ppf = lazy (Format.make_formatter (fun _ _ _ -> ()) (fun () -> ()))
+
+let unformat (f : Format.formatter -> unit) : string =
+  let buf = Buffer.create 1024 in
+  let ppf = unstyle_formatter (Format.formatter_of_buffer buf) in
+  Format.pp_set_margin ppf max_int;
+  (* We won't print newlines anyways, but better not have them in the first
+     place (this wouldn't remove cuts in a vbox for example) *)
+  let out_funs = Format.pp_get_formatter_out_functions ppf () in
+  Format.pp_set_formatter_out_functions ppf
+    {
+      out_funs with
+      Format.out_newline = (fun () -> out_funs.out_string " " 0 1);
+      Format.out_indent = (fun _ -> ());
+    };
+  f ppf;
+  Format.pp_print_flush ppf ();
+  Buffer.contents buf
+
+(**{2 Message types and output helpers *)
 
-(**{2 Printers}*)
-
-(** All the printers below print their argument after the correct marker *)
-
-let debug_format (format : ('a, Format.formatter, unit) format) =
-  if !Cli.debug_flag then
-    Format.printf ("%a@[<hov>" ^^ format ^^ "@]@.") debug_marker ()
-  else Format.ifprintf Format.std_formatter format
-
-let error_format (format : ('a, Format.formatter, unit) format) =
-  Format.print_flush ();
-  (* Flushes previous warnings *)
-  Format.printf ("%a" ^^ format ^^ "\n%!") error_marker ()
+type content_type = Error | Warning | Debug | Log | Result
 
-let warning_format format =
-  if !Cli.disable_warnings_flag then Format.ifprintf Format.std_formatter format
-  else Format.printf ("%a" ^^ format ^^ "\n%!") warning_marker ()
+let get_ppf = function
+  | Result -> Lazy.force std_ppf
```

Comment:
Why `Lazy` ?

Reply:
There should be this comment in `messages.mli` that describes how we expect these side-effects to work I think :)